Sources/SecureXPC/Client/XPCClient.swift

```diff
@@ -126,16 +126,36 @@ public class XPCClient {
         XPCMachClient(serviceName: machServiceName)
     }
 
-	// MARK: Implementation
+	public static func forEndpoint(_ endpoint: XPCServerEndpoint) -> XPCClient {
+        let connection = xpc_connection_create_from_endpoint(endpoint.endpoint)
 
-    internal let serviceName: String
+        xpc_connection_set_event_handler(connection, { (event: xpc_object_t) in
+            // A block *must* be set as the handler, even though this block does nothing.
+            // If it were not set, a crash would occur upon calling xpc_connection_resume.
+            xpc_connection_cancel(event)
+        })
+        xpc_connection_resume(connection)
+
+        switch endpoint.kind {
+        case .xpcServiceClient:
+            return XPCServiceClient(serviceName: endpoint.serviceName, connection: connection)
+        case .machServiceClient:
+            return XPCMachClient(serviceName: endpoint.serviceName!, connection: connection)
+        }
+    }
+
+    // MARK: Implementation
+
+    internal let serviceName: String?
+    private var connection: xpc_connection_t? = nil
     
     /// Creates a client which will attempt to send messages to the specified mach service.
     ///
     /// - Parameters:
     ///   - serviceName: The name of the XPC service; no validation is performed on this.
```

Comment:
Ah yes, this doc comment is a mess.

> Additionally both parameters can be nil but if both are is that actually valid?

Yes.

The `serviceName` can be `nil` for anonymous services. For regular (named) XPC services, and Mach services, it needs to be non-nil. Perhaps `serviceName` should be an abstract property, and the subclasses can store their own name variable (which can be `non-nil` for mach services.

When creating a client from an `XPCEndpoint`, the connection is already created, and so the client just uses that. In other cases, `connection` will be `nil`, and the client will create a new connection for itself from the service name.

Reply:
Actually looking at this in more details, looks like there's no need for `XPCClient` to have a `serviceName` instance variable. This can purely be an internal implementation detail of the `XPCMachClient` and `XPCServiceClient` subclasses and a future `XPCAnonymousClient` (or whatever it's called) subclasses wouldn't have one.

Similarly there doesn't appear to be any benefit to passing the connection into the initializer as it'll get created (or return an existing one for an `XPCEndpoint`) as part of the `createConnection()` function. Is that right?

Reply:
> Actually looking at this in more details, looks like there's no need for `XPCClient` to have a `serviceName` instance variable. This can purely be an internal implementation detail of the `XPCMachClient` and `XPCServiceClient` subclasses and a future `XPCAnonymousClient` (or whatever it's called) subclasses wouldn't have one.

Errrr I disagree. NSXPCConnection and the C APIs both take this approach, where connections are just these magical opaque things with very little exposed API. They happen to print their service name (if known) in their `description`, but they're still annoying to deal with, because e.g. their name isn't visible from the object view in the debugger.

I think having the service name be exposed is easy enough that we should have it. It's just a nice convenience for development. (not that anybody should ever need to like, `switch` over it, or anything like that)

> Similarly there doesn't appear to be any benefit to passing the connection into the initializer as it'll get created (or return an existing one for an XPCEndpoint) as part of the createConnection() function. Is that right?

How will "or return an existing one for an XPCEndpoint" work if the connection isn't passed in?

Reply:
I still have to update the docs here (and elsewhere, probably), but could yo uplease look over my last comment?

Reply:
> NSXPCConnection and the C APIs both take this approach, where connections are just these magical opaque things with very little exposed API. They happen to print their service name (if known) in their description, but they're still annoying to deal with, because e.g. their name isn't visible from the object view in the debugger.

Hmm, I think we just have a difference of opinion when it comes to API design here and I think Apple took the right approach. As far as I see it, this is why `CustomDebugStringConvertible` exists. Unless I'm mistaken the value of `debugDescription` will be visible in the debugger?

> How will "or return an existing one for an XPCEndpoint" work if the connection isn't passed in?

It'd get passed into the initializer of a concrete subclass.